conformance/utils/kubernetes/helpers_test.go

```diff
@@ -0,0 +1,160 @@
+						Reason: string(v1beta1.ListenerReasonRefNotPermitted),
+					}},
+				},
+			},
+			actual: []v1beta1.ListenerStatus{
+				{
+					Name: v1beta1.SectionName("https"),
+					SupportedKinds: []v1beta1.RouteGroupKind{{
+						Group: (*v1beta1.Group)(&v1beta1.GroupVersion.Group),
+						Kind:  v1beta1.Kind("HTTPRoute"),
+					}},
+					Conditions: []metav1.Condition{{
+						Type:   string(v1beta1.ListenerConditionResolvedRefs),
+						Status: metav1.ConditionFalse,
+						Reason: string(v1beta1.ListenerReasonRefNotPermitted),
+					}},
+				},
+			},
+			want: true,
+		},
+		{
+			name: "expected kind not found in actual",
+			expected: []v1beta1.ListenerStatus{
+				{
+					Name: v1beta1.SectionName("https"),
+					SupportedKinds: []v1beta1.RouteGroupKind{{
+						Group: (*v1beta1.Group)(&v1beta1.GroupVersion.Group),
+						Kind:  v1beta1.Kind("HTTPRoute"),
+					}},
+					Conditions: []metav1.Condition{{
+						Type:   string(v1beta1.ListenerConditionResolvedRefs),
+						Status: metav1.ConditionFalse,
+						Reason: string(v1beta1.ListenerReasonRefNotPermitted),
+					}},
+				},
+			},
+			actual: []v1beta1.ListenerStatus{
+				{
+					Name: v1beta1.SectionName("https"),
+					SupportedKinds: []v1beta1.RouteGroupKind{
+						{
+							Group: (*v1beta1.Group)(&v1alpha2.GroupVersion.Group),
+							Kind:  v1beta1.Kind("GRPCRoute"),
+						},
+					},
+					Conditions: []metav1.Condition{{
+						Type:   string(v1beta1.ListenerConditionResolvedRefs),
+						Status: metav1.ConditionFalse,
+						Reason: string(v1beta1.ListenerReasonRefNotPermitted),
+					}},
+				},
+			},
+			want: false,
+		},
+		{
+			name: "expected is a subset of actual",
+			expected: []v1beta1.ListenerStatus{
+				{
+					Name: v1beta1.SectionName("https"),
+					SupportedKinds: []v1beta1.RouteGroupKind{{
+						Group: (*v1beta1.Group)(&v1beta1.GroupVersion.Group),
+						Kind:  v1beta1.Kind("HTTPRoute"),
+					}},
+					Conditions: []metav1.Condition{{
+						Type:   string(v1beta1.ListenerConditionResolvedRefs),
+						Status: metav1.ConditionFalse,
+						Reason: string(v1beta1.ListenerReasonRefNotPermitted),
+					}},
+				},
+			},
+			actual: []v1beta1.ListenerStatus{
+				{
+					Name: v1beta1.SectionName("https"),
+					SupportedKinds: []v1beta1.RouteGroupKind{
+						{
+							Group: (*v1beta1.Group)(&v1alpha2.GroupVersion.Group),
+							Kind:  v1beta1.Kind("GRPCRoute"),
+						},
+						{
+							Group: (*v1beta1.Group)(&v1beta1.GroupVersion.Group),
+							Kind:  v1beta1.Kind("HTTPRoute"),
+						},
+					},
+					Conditions: []metav1.Condition{{
+						Type:   string(v1beta1.ListenerConditionResolvedRefs),
+						Status: metav1.ConditionFalse,
+						Reason: string(v1beta1.ListenerReasonRefNotPermitted),
+					}},
+				},
+			},
+			want: true,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			assert.Equal(t, test.want, listenersMatch(t, test.expected, test.actual))
```

Comment:
One thing I noticed is that because we only test for `bool` return values, and we don't have a mechanism here to verify the logging output our checking of results is a bit limited. That said however, this isn't something I think we need to resolve right here and now because it would grow the scope of this PR considerably to refactor this now, and this is blocking the `v0.6.1` release. As such, as per @mikemorris' comments throughout, I'll make a follow-up issue to do some refactoring in here in a later iteration.

Reply:
https://github.com/kubernetes-sigs/gateway-api/issues/1708 was created to resolve this comment :+1: